File: HiTaxi_Java/app/src/main/java/com/huawei/hitaxicodelab_java/ui/destination/DestinationFragment.java
/*
**********************************************************************************
|                                                                                |
| Copyright 2021. Huawei Technologies Co., Ltd. All rights reserved.             |
|                                                                                |
| Licensed under the Apache License, Version 2.0 (the "License");                |
| you may not use this file except in compliance with the License.               |
| You may obtain a copy of the License at                                        |
|                                                                                |
| http://www.apache.org/licenses/LICENSE-2.0                                     |
|                                                                                |
| Unless required by applicable law or agreed to in writing, software            |
| distributed under the License is distributed on an "AS IS" BASIS,              |
| WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.       |
| See the License for the specific language governing permissions and            |
| limitations under the License.                                                 |
|                                                                                |
**********************************************************************************
*/

package com.huawei.hitaxicodelab_java.ui.destination;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;

import com.huawei.hitaxicodelab_java.BuildConfig;
import com.huawei.hitaxicodelab_java.R;
import com.huawei.hitaxicodelab_java.databinding.FragmentDestinationBinding;
import com.huawei.hitaxicodelab_java.ui.base.BaseFragment;
import com.huawei.hitaxicodelab_java.ui.home.HomeViewModel;
import com.huawei.hms.maps.CameraUpdateFactory;
import com.huawei.hms.maps.HuaweiMap;
import com.huawei.hms.maps.OnMapReadyCallback;
import com.huawei.hms.maps.model.LatLng;
import com.huawei.hms.site.api.model.Site;
import com.huawei.hms.site.widget.SearchIntent;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class DestinationFragment extends BaseFragment<HomeViewModel, FragmentDestinationBinding>
        implements OnMapReadyCallback {

    HuaweiMap huaweiMap;
    SearchIntent searchIntent;
    HomeViewModel homeViewModel;

    String tempStartingAddress;
    LatLng tempStartingLocation;
    String tempDestinationAddress;
    LatLng tempDestinationLocation;

    @NotNull
    @Override
    public HomeViewModel getViewModel() {
        return homeViewModel;
    }

    @NotNull
    @Override
    public FragmentDestinationBinding getFragmentViewBinding(
            @NotNull LayoutInflater inflater,
            @Nullable ViewGroup container) {
        return FragmentDestinationBinding.inflate(inflater, container, false);
    }

    @Override
    public void onViewCreated(@NotNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initSearchIntent();
    }

    @Override
    public void setupListeners() {
        super.setupListeners();
        getFragmentViewBinding().searchLayout.setOnClickListener(view -> showSiteSearchFragment());

        getFragmentViewBinding().myLocationButton.setOnClickListener(view -> focusUserCurrentLocation());

        getFragmentViewBinding().setDestinationButton.setOnClickListener(view -> {
            if (DestinationFragmentArgs.fromBundle(getArguments()).getIsFromDestination()) {
                if (Boolean.FALSE.equals(homeViewModel.getIsDestinationSelectedBefore().getValue())) {
                    homeViewModel.setIsDestinationSelectedBefore(true);
                    homeViewModel.setStartingAddress(getString(R.string.your_location));
                    if (homeViewModel.getLastKnownLocation().getValue() != null) {
                        homeViewModel.setStartingLocation(homeViewModel.getLastKnownLocation().getValue());
                    }
                }
                LatLng location = tempDestinationLocation;
                String address = tempDestinationAddress;
                if (location != null && address != null) {
                    homeViewModel.setDestinationLocation(location);
                    homeViewModel.setDestinationAddress(address);
                    NavHostFragment.findNavController(this).navigate(R.id.action_destinationFragment_to_homeFragment);
                }
            } else {
                LatLng location = tempStartingLocation;
                String address = tempStartingAddress;
                if (location != null && address != null) {
                    homeViewModel.setStartingLocation(location);
                    homeViewModel.setStartingAddress(address);
                    NavHostFragment.findNavController(this).navigate(R.id.action_destinationFragment_to_homeFragment);
                }
            }
        });
    }

    @Override
    public void setupUi() {
        super.setupUi();
        homeViewModel = new ViewModelProvider(requireActivity()).get(HomeViewModel.class);
        getFragmentViewBinding().huaweiMap.onCreate(null);
        getFragmentViewBinding().huaweiMap.getMapAsync(this);
        changeLocationPinColor();
    }

    @Override
    public void onMapReady(HuaweiMap map) {
        huaweiMap = map;

        huaweiMap.getUiSettings().setZoomControlsEnabled(false);
        huaweiMap.setOnCameraIdleListener(() -> {
            double latitude = huaweiMap.getCameraPosition().target.latitude;
            double longitude = huaweiMap.getCameraPosition().target.longitude;
            LatLng latLng = new LatLng(latitude, longitude);
            saveAndShowAddress(latLng);
        });

        if (DestinationFragmentArgs.fromBundle(getArguments()).getIsFromDestination()) {
            if (homeViewModel.getDestinationLocation().getValue() == null) {
                focusUserCurrentLocation();
            } else {
                LatLng location = homeViewModel.getDestinationLocation().getValue();
                huaweiMap.moveCamera(CameraUpdateFactory.newLatLngZoom(location, 18f));
            }
        } else {
            if (homeViewModel.getStartingLocation().getValue() == null) {
                focusUserCurrentLocation();
            } else {
                LatLng location = homeViewModel.getStartingLocation().getValue();
                huaweiMap.moveCamera(CameraUpdateFactory.newLatLngZoom(location, 18f));
            }
        }
    }

    private void initSearchIntent() {
        searchIntent = new SearchIntent();
        try {
            final String encodedPath = URLEncoder.encode(BuildConfig.API_KEY, "UTF-8");
            searchIntent.setApiKey(encodedPath);
        }catch(UnsupportedEncodingException ec) {
            Log.d("TAG", ec.getLocalizedMessage());
        }
    }

    private void changeLocationPinColor() {
        if (!DestinationFragmentArgs.fromBundle(getArguments()).getIsFromDestination()) {
            getFragmentViewBinding().locationPinImageView.setImageDrawable(
                    ContextCompat.getDrawable(
                            requireContext(),
                            R.drawable.ic_start_location_pin));
        }
    }

    private void focusUserCurrentLocation() {
        if (homeViewModel.getLastKnownLocation().getValue() != null) {
            huaweiMap.moveCamera(CameraUpdateFactory.newLatLngZoom(homeViewModel.getLastKnownLocation().getValue(), 18f));
        }
    }

    private void showSiteSearchFragment() {
        Intent intent = searchIntent.getIntent(requireActivity());
        startActivityForResult(intent, SearchIntent.SEARCH_REQUEST_CODE);
    }

    @Override
    @Deprecated
    public void onActivityResult(int requestCode, int resultCode, @androidx.annotation.Nullable @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (SearchIntent.SEARCH_REQUEST_CODE == requestCode && SearchIntent.isSuccess(resultCode)) {
            Site site = searchIntent.getSiteFromIntent(data);
            LatLng latLng = new LatLng(site.location.lat, site.location.lng);
            huaweiMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 18f));
            saveAndShowAddress(latLng);
        }
    }

    private void saveAndShowAddress(LatLng latLng) {
        getFragmentViewBinding().searchAddressText.setText(homeViewModel.getAddress(latLng));

        if (DestinationFragmentArgs.fromBundle(getArguments()).getIsFromDestination()) {
            tempDestinationAddress = homeViewModel.getAddress(latLng);
            tempDestinationLocation = latLng;
        } else {
            tempStartingAddress = homeViewModel.getAddress(latLng);
            tempStartingLocation = latLng;
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        getFragmentViewBinding().huaweiMap.onStart();
    }

    @Override
    public void onStop() {
        super.onStop();
        getFragmentViewBinding().huaweiMap.onStop();
    }

    @Override
    public void onResume() {
        super.onResume();
        getFragmentViewBinding().huaweiMap.onResume();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        getFragmentViewBinding().huaweiMap.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        getFragmentViewBinding().huaweiMap.onLowMemory();
    }
}

File: HiTaxi_Kotlin/app/src/main/java/com/huawei/hitaxicodelab/ui/destination/DestinationFragment.kt
/*
**********************************************************************************
|                                                                                |
| Copyright 2021. Huawei Technologies Co., Ltd. All rights reserved.             |
|                                                                                |
| Licensed under the Apache License, Version 2.0 (the "License");                |
| you may not use this file except in compliance with the License.               |
| You may obtain a copy of the License at                                        |
|                                                                                |
| http://www.apache.org/licenses/LICENSE-2.0                                     |
|                                                                                |
| Unless required by applicable law or agreed to in writing, software            |
| distributed under the License is distributed on an "AS IS" BASIS,              |
| WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.       |
| See the License for the specific language governing permissions and            |
| limitations under the License.                                                 |
|                                                                                |
**********************************************************************************
*/
package com.huawei.hitaxicodelab.ui.destination

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import com.huawei.hitaxicodelab.BuildConfig
import com.huawei.hitaxicodelab.R
import com.huawei.hitaxicodelab.databinding.FragmentDestinationBinding
import com.huawei.hitaxicodelab.ui.base.BaseFragment
import com.huawei.hitaxicodelab.ui.home.HomeViewModel
import com.huawei.hms.maps.CameraUpdateFactory
import com.huawei.hms.maps.HuaweiMap
import com.huawei.hms.maps.OnMapReadyCallback
import com.huawei.hms.maps.model.LatLng
import com.huawei.hms.site.widget.SearchIntent
import dagger.hilt.android.AndroidEntryPoint
import java.net.URLEncoder

@AndroidEntryPoint
class DestinationFragment : BaseFragment<HomeViewModel, FragmentDestinationBinding>(),
    OnMapReadyCallback {

    private val homeViewModel: HomeViewModel by activityViewModels()

    override fun getViewModel(): HomeViewModel = homeViewModel

    private lateinit var searchIntent: SearchIntent
    private lateinit var huaweiMap: HuaweiMap

    private val args: DestinationFragmentArgs by navArgs()

    var tempStartingAddress: String? = null
    var tempStartingLocation: LatLng? = null
    var tempDestinationAddress: String? = null
    var tempDestinationLocation: LatLng? = null

    override fun getFragmentViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentDestinationBinding {
        return FragmentDestinationBinding.inflate(inflater, container, false)
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initSearchIntent()
    }

    override fun setupUi() {
        super.setupUi()
        fragmentViewBinding.huaweiMap.apply {
            onCreate(null)
            getMapAsync(this@DestinationFragment)
        }
        changeLocationPinColor()
    }

    override fun setupListeners() {
        fragmentViewBinding.searchLayout.setOnClickListener {
            showSiteSearchFragment()
        }

        fragmentViewBinding.myLocationButton.setOnClickListener {
            focusUserCurrentLocation()
        }

        fragmentViewBinding.setDestinationButton.setOnClickListener {

            if (args.isFromDestination) {
                if (homeViewModel.isDestinationSelectedBefore.value == false) {
                    homeViewModel.setIsDestinationSelectedBefore(true)
                    homeViewModel.setStartingAddress(getString(R.string.your_location))
                    homeViewModel.lastKnownLocation.value?.let {
                        homeViewModel.setStartingLocation(it)
                    }
                }
                val location = tempDestinationLocation
                val address = tempDestinationAddress
                if (location != null && address != null) {
                    homeViewModel.setDestinationLocation(location)
                    homeViewModel.setDestinationAddress(address)
                    findNavController().navigate(R.id.action_destinationFragment_to_homeFragment)
                }
            } else {
                val location = tempStartingLocation
                val address = tempStartingAddress
                if (location != null && address != null) {
                    homeViewModel.setStartingLocation(location)
                    homeViewModel.setStartingAddress(address)
                    findNavController().navigate(R.id.action_destinationFragment_to_homeFragment)
                }
            }
        }
    }


    private fun saveAndShowAddress(latLng: LatLng) {
        fragmentViewBinding.searchAddressText.text = homeViewModel.getAddress(latLng)

        if (args.isFromDestination) {
            tempDestinationAddress = homeViewModel.getAddress(latLng)
            tempDestinationLocation = latLng
        } else {
            tempStartingAddress = homeViewModel.getAddress(latLng)
            tempStartingLocation = latLng
        }
    }


    override fun onMapReady(map: HuaweiMap) {
        huaweiMap = map

        huaweiMap.apply {
            uiSettings.isZoomControlsEnabled = false
            setOnCameraIdleListener {
                val latitude = huaweiMap.cameraPosition.target.latitude
                val longitude = huaweiMap.cameraPosition.target.longitude
                val latLng = LatLng(latitude, longitude)
                saveAndShowAddress(latLng)
            }
        }


        if (args.isFromDestination) {
            if (homeViewModel.destinationLocation.value == null) {
                focusUserCurrentLocation()
            } else {
                val location = homeViewModel.destinationLocation.value
                huaweiMap.moveCamera(CameraUpdateFactory.newLatLngZoom(location, 18f))
            }
        } else {
            if (homeViewModel.startingLocation.value == null) {
                focusUserCurrentLocation()
            } else {
                val location = homeViewModel.startingLocation.value
                huaweiMap.moveCamera(CameraUpdateFactory.newLatLngZoom(location, 18f))
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (SearchIntent.SEARCH_REQUEST_CODE == requestCode) {
            if (SearchIntent.isSuccess(resultCode)) {
                val site = searchIntent.getSiteFromIntent(data)
                val latLng = LatLng(site.location.lat, site.location.lng)
                huaweiMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 18f))
                saveAndShowAddress(latLng)
            }
        }
    }

    private fun initSearchIntent() {
        searchIntent = SearchIntent()
        searchIntent.setApiKey(URLEncoder.encode(BuildConfig.API_KEY, "UTF-8"))
    }

    private fun showSiteSearchFragment() {
        val intent = searchIntent.getIntent(requireActivity())
        startActivityForResult(intent, SearchIntent.SEARCH_REQUEST_CODE)
    }


    private fun changeLocationPinColor() {
        if (args.isFromDestination.not()) {
            fragmentViewBinding.locationPinImageView.setImageDrawable(
                ContextCompat.getDrawable(
                    requireContext(),
                    R.drawable.ic_start_location_pin
                )
            )
        }
    }

    private fun focusUserCurrentLocation() {
        homeViewModel.lastKnownLocation.value?.let {
            huaweiMap.moveCamera(CameraUpdateFactory.newLatLngZoom(it, 18f))
        }
    }

    override fun onStart() {
        super.onStart()
        fragmentViewBinding.huaweiMap.onStart()
    }

    override fun onStop() {
        super.onStop()
        fragmentViewBinding.huaweiMap.onStop()
    }

    override fun onResume() {
        super.onResume()
        fragmentViewBinding.huaweiMap.onResume()
    }

    override fun onDestroy() {
        super.onDestroy()
        fragmentViewBinding.huaweiMap.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        fragmentViewBinding.huaweiMap.onLowMemory()
    }
}